components/scoreHandler.ts

```diff
@@ -1019,6 +1020,15 @@ export async function missionEnd(
         }
     }
 
+    /**
+     * Clears inventory if new unlockables have been unlocked, so the inventory is recreated with them instead of fetching from the cache
+     *
+     * Im tempted to avoid doing this if the mission type is evergreen, because I dont think we need to clear inventory for freelancer, but would need to test (would save some cpu usage)
+     */
+    if (drops.length) {
+        clearInventoryFor(req.jwt.unique_name)
+    }
+
```

Comment:
Putting the performance issue aside, this way of doing unlocks will only grant unlockables at mission end screens. However, they should be granted to the user as soon as the user earns enough xp to unlock them, as per the behavior on official. Please take a look into the `Drop`s of challenges and mastery. 

A more ideal implementation might be:
- When booting, construct the inventory based on the user's entitlements, mastery progression, and challenge completion.
- Whenever a challenge is completed, check its `Drop` (if any) and grant it to the user.
- Whenever XP is gained, check the mastery level and grant corresponding `Drop`s to the user.